cmd/cronosd/cmd/fix-unlucky-tx.go

```diff
@@ -443,33 +445,94 @@ func logTnxHash(txConfig client.TxConfig, result *abci.TxResult, action string)
 	}
 }
 
-func (db *tmDB) PatchFromImport(txConfig client.TxConfig, reader io.Reader) error {
+func (db *tmDB) PatchFromImport(txConfig client.TxConfig, reader io.Reader, concurrency int) (returnErr error) {
 	maxItemSize := int(64e6)
 	protoReader := protoio.NewDelimitedReader(reader, maxItemSize)
 	var err error
+
+	type importData struct {
+		Result      *abci.TxResult
+		BlockResult *tmstate.ABCIResponses
+	}
+	chData := make(chan importData)
+	chError := make(chan error, concurrency+1)
+	var wg sync.WaitGroup
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	for i := 0; i < concurrency; i++ {
+		wg.Add(1)
+		go func(wg *sync.WaitGroup) {
+			defer wg.Done()
+			for {
+				select {
+				case data, ok := <-chData:
+					if !ok {
+						return
+					}
+					if err := db.txIndexer.Index(data.Result); err != nil {
+						chError <- err
+						cancel()
+						return
+					}
+					if err := db.stateStore.SaveABCIResponses(data.Result.Height, data.BlockResult); err != nil {
+						chError <- err
+						cancel()
+						return
+					}
+				case <-ctx.Done():
+					return
+				}
+			}
+		}(&wg)
+	}
+
+	go func() {
+		for returnErr = range chError {
+			break
+		}
+	}()
+
 	for {
-		res := abci.TxResult{}
-		_, err = protoReader.ReadMsg(&res)
+		data := importData{
+			Result:      new(abci.TxResult),
+			BlockResult: new(tmstate.ABCIResponses),
+		}
+		_, err = protoReader.ReadMsg(data.Result)
```

Comment:
db update is fast, bottleneck comes from read msg from single resource, patch from whole 32G data take 1h

Reply:
Is the 1h after this PR or before?

Reply:
only 10mins faster, seems tm [reader](https://github.com/tendermint/tendermint/blob/17f4ea3680de0accba030c811ab7d08a451a2b3d/libs/protoio/reader.go#L47) still read proto one by one, need read like `3 x x x 4 y y y y`, we pass data in between goroutine unmarshal to be real concurrent

Reply:
I tried save abci by [batch](https://github.com/mmsqe/tendermint/commit/44c006ba7c8f46eaf2087b0bad5b0535a4682311), which save us 30min